### ISSUE TITLE
[dynamo] dynamo fails to compile a correct dynamic graph and lead to unexpected recompiles

### ISSUE THREAD

**Issue by yyp0** [pytorch/pytorch#146168]
### 🐛 Describe the bug

Dynamo is expected to compile a dynamic graph when some specific sizes changes. However, in our case, it seems dynamo fails to do that. Graph in the first iteration:
```
class GraphModule(torch.nn.Module):
    def forward(self, L_inputs_ : list, ...):
        getitem: "bf16[**4090**, 1, 4096]" = l_inputs_[0]
        view: "bf16[4090, 4096]" = torch.ops.aten.view.default(getitem, [**4090**, 4096])
```
Graph in the second iteration:
```
class GraphModule(torch.nn.Module):
    def forward(self, L_inputs_ : list, **L_sizes_4_: "Sym(4089)"**,  ...):
         getitem: "bf16[**4089**, 1, 4096]" = l_inputs_[0]
         view: "bf16[4089, 4096]" = torch.ops.aten.view.default(getitem, [**l_sizes_4_**, 4096]);  l_sizes_4_ = None
```
It seems that the scalar(L_sizes_4_) and the dynamic dim (4090 vs 4089) are folded, and the following guards are added, leading to guard check fails and recompiles:
```
L['sizes'][4] == 4089  # view = torch.ops.aten.view.default(getitem, [getitem_63, 4096]);  getitem_63 = None  # <eval_with_key>.3:244 in forward (_refs/__init__.py:3755 in _reshape_view_helper)
```
Do you know why  the `L_sizes_4_` and `getitem` are treated as constants when the dim changes? By the way, I have enabled compiled_autograd in the current case and just trace the backward graph. I'm not sure whether the compiled_autograd has any side effects on dynamo's dynamic feature. 

cc @chauhang @penguinwu @ezyang @bobrenjc93 @xmfan @yf225 

### Versions

pytorch 2.6

**Comment by xmfan** [comment]
This might be the same issue as https://github.com/pytorch/pytorch/issues/113129#issuecomment-2322794167.

Until it's fixed, there's 2 ways to reduce the number of recompiles:
1. if you're enabling via the config `torch._dynamo.compiled_autograd=True`, you can consider `torch.compile(dynamic=True)`, but this may cause a lot of more obscure dynamic shapes issues documented here: https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit?tab=t.0#heading=h.fh8zzonyw8ng.
2. there's a private API to allow compiled autograd to always trace the autograd graph as dynamic (without applying it to the torch.compile calls): `torch._dynamo.compiled_autograd._enable(, dynamic=True)`. I'm thinking of changing this to default on

**Comment by yyp0** [comment]
> This might be the same issue as [#113129 (comment)](https://github.com/pytorch/pytorch/issues/113129#issuecomment-2322794167).
> 
> Until it's fixed, there's 2 ways to reduce the number of recompiles:
> 
> 1. if you're enabling via the config `torch._dynamo.compiled_autograd=True`, you can consider `torch.compile(dynamic=True)`, but this may cause a lot of more obscure dynamic shapes issues documented here: https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit?tab=t.0#heading=h.fh8zzonyw8ng.
> 2. there's a private API to allow compiled autograd to always trace the autograd graph as dynamic (without applying it to the torch.compile calls): `torch._dynamo.compiled_autograd._enable(, dynamic=True)`. I'm thinking of changing this to default on

Thanks for your reply! But we have maintained an inner repo which is not forked from torch26 released versions, and do not have dynamic arg in the api `torch._dynamo.compiled_autograd.enable`. It is not very easy for us to upgrade torch version temporarily. Does the usage `with torch._dynamo.compiled_autograd.enable(torch.compile(backend=my_compiler, fullgraph=True, dynamic=True)):` have the same function with the mentioned two ways? Or is there any method to hack the issue, like preventing constant folding?

**Comment by xmfan** [comment]
If you want to keep it as a toggle, you could patch in the implementation here: https://github.com/pytorch/pytorch/pull/141153. 

A simpler patch by defaulting the CA compilation to dynamic is to change this constructor:

https://github.com/pytorch/pytorch/blob/d2a2b9f8a74218d685373e3b942e591dcc8c5c0a/torch/csrc/dynamo/compiled_autograd.h#L230-L234

to instead always
```python
  AutogradCompilerCall(SizeInput::DynType default_dyn_type)
      : active_node_call_idx(std::nullopt),
        tensor_args(active_node_call_idx),
        lifted_ivalue_args(active_node_call_idx),
        default_dyn_type(SizeInput::DynType::DYNAMIC) {}
```

**Comment by yyp0** [comment]
> If you want to keep it as a toggle, you could patch in the implementation here: [#141153](https://github.com/pytorch/pytorch/pull/141153).
> 
> A simpler patch by defaulting the CA compilation to dynamic is to change this constructor:
> 
> [pytorch/torch/csrc/dynamo/compiled_autograd.h](https://github.com/pytorch/pytorch/blob/d2a2b9f8a74218d685373e3b942e591dcc8c5c0a/torch/csrc/dynamo/compiled_autograd.h#L230-L234)
> 
> Lines 230 to 234 in [d2a2b9f](/pytorch/pytorch/commit/d2a2b9f8a74218d685373e3b942e591dcc8c5c0a)
> 
>  AutogradCompilerCall(SizeInput::DynType default_dyn_type) 
>      : active_node_call_idx(std::nullopt), 
>        tensor_args(active_node_call_idx), 
>        lifted_ivalue_args(active_node_call_idx), 
>        default_dyn_type(default_dyn_type) {} 
> to instead always
> 
>   AutogradCompilerCall(SizeInput::DynType default_dyn_type)
>       : active_node_call_idx(std::nullopt),
>         tensor_args(active_node_call_idx),
>         lifted_ivalue_args(active_node_call_idx),
>         default_dyn_type(SizeInput::DynType::DYNAMIC) {}

Thanks! I have upgrade the pytorch, and enbale compiled_autograd(torch.compile(..., dynamic=True), dynamic=True). Based on it, I have note that the compiled_autograd::begin_capture is only lanuched once, which, i think, indicates that compiled autograd returns a fully dynamic graph and just compile once. However, dynamo still keeps recompiles when giving inputs with different seq len. Do you have some other advice to prevent dynamo's recompile? It is okay for us to get a fully dynamic graph, for we only do some high level optimization and keep the graph eager run.

**Comment by xmfan** [comment]
You can check the dynamo recompilation reasons with either TORCH_LOGS="recompiles". There's also a UI version via tlparse: https://github.com/pytorch/tlparse
